### PR TITLE
oauthdevice: Rename to `auth`; simplify interface

### PR DESCRIPTION
**Pull request by minor-fixes**
This change simplifies the `oauthdevice.Authenticator` interface from
the two-step `FetchCode` + `FetchToken` to a combined single method
`Authenticate` that essentially chains the two code paths. The former
interface may over-fit the device code auth flow, whereas the new
interface may allow for non-device-code-flows to be implemented in the
future.

Since this change causes the interface to no longer be
device-code-specific, the former package name no longer makes sense.
This change results in the following renames:

`package oauthdevice` -> `package auth`
`oauthdevice.Authenticator` -> `auth.Backend`
`oauthdevice.Auth` -> `auth.DeviceCode`
`oauthdevice.NewAuth` -> `auth.NewDeviceCode`

This change moves the browser-opening step from `main` into
`auth.DeviceCode` as a consequence (which more accurately reflects
the reality that not all auth flows involve opening a browser). Tests
for main are simpler but tests for `auth.DeviceCode` are more complex as
a result.

Tests for `auth.DeviceCode` mock the underlying transport to perform
validation on the HTTP requests and responses, bringing some of the
`oauth2` package logic into the test. This results in slight
over-testing, but may allow us to more easily validate `engflow_auth`
against actual HTTP responses from our oauth endpoint in the future.

The first dependency on `github.com/stretchr/testify/mock` requires us to add
`github.com/stretchr/objx` as a dependency, which requires a [security approval](https://docs.google.com/spreadsheets/d/1uTKuur4qbow2tRw7qvRfmReSJWjzhfpKtUM3Zlh7Ppg/edit?gid=0#gid=0)

### CODE REVIEW

**Comment by minor-fixes** [comment]
note to reviewers: Don't feel especially strong about this - I just have a sneaking suspicion that the original interface is too narrow to be futureproof, and thought that the best time to change it was sooner rather than later

File: internal/auth/authenticator.go
Comment (rogerhu) [inline internal/auth/authenticator.go:73]: btw, is this a bug now?

Comment (minor-fixes) [reply internal/auth/authenticator.go:73]: Dunno - do we have any non-oauth-aware backends running, or have we enabled the feature everywhere?

Comment (rogerhu) [reply internal/auth/authenticator.go:73]: It got merged in https://github.com/EngFlow/engflow/pull/22227.

**Comment by rogerhu** [comment]
Seems fine to me though it moves the control flow for opening browser into the interface. 

@jayconrod should comment about Golang hygeines.

File: internal/auth/authenticator.go
Comment (minor-fixes) [reply internal/auth/authenticator.go:73]: The check should stay until *no* backends exist that can return this kind of response

Comment (jayconrod) [reply internal/auth/authenticator.go:73]: We could rephrase to say that older versions of EngFlow redirected unknown API requests to a login page that eventually responded with 200 and HTML. That shouldn't happen anymore, but we can still guard against it. The new behavior is to return 404 instead when OAuth2 device flow is disabled.

File: internal/auth/authenticator.go
Comment (jayconrod) [inline internal/auth/authenticator.go:32]: Exported definitions should have doc comments if their purpose isn't entirely clear from their names. Doubly so for types.

File: internal/auth/authenticator.go
Comment (jayconrod) [inline internal/auth/authenticator.go:40]: It doesn't seem like this can be initialized by the constructor. What's it for?

File: internal/auth/authenticator.go
Comment (jayconrod) [inline internal/auth/authenticator.go:52]: What's happening here?

File: internal/auth/authenticator_test.go
Comment (jayconrod) [inline internal/auth/authenticator_test.go:32]: Unexport: this can't be used by other packages.

File: go.mod
Comment (jayconrod) [inline go.mod:23]: Is this covered by security review already?

Comment (minor-fixes) [reply go.mod:23]: Just checked - it's not :/

I created https://docs.google.com/spreadsheets/d/1uTKuur4qbow2tRw7qvRfmReSJWjzhfpKtUM3Zlh7Ppg/edit?gid=0#gid=0; will update the PR desc and add security

File: internal/auth/authenticator_test.go
Comment (minor-fixes) [reply internal/auth/authenticator_test.go]: Done!

File: internal/auth/authenticator.go
Comment (minor-fixes) [reply internal/auth/authenticator.go:69]: Added a comment

File: internal/auth/authenticator.go
Comment (minor-fixes) [reply internal/auth/authenticator.go:50]: Added a comment here, and explicitly set it to nil in the constructor so that it looks more intentional.

File: internal/auth/authenticator.go
Comment (minor-fixes) [reply internal/auth/authenticator.go]: Ok, rephrased

File: internal/auth/authenticator.go
Comment (minor-fixes) [reply internal/auth/authenticator.go:34]: Added comments for `Backend` and `DeviceCode`; I think `NewDeviceCode` and `DeviceCode.Authenticate` are fairly obvious from these docs; WDYT?

**Review by jayconrod — APPROVED** [review]

**Comment by jayconrod** [comment]
(LGTM by me, but still needs security approval for the new dependency).

**Pull request closed without merging**
